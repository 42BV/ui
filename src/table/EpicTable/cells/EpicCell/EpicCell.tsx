import React, { useRef } from 'react';
import classNames from 'classnames';

type Props = {
  /**
   * The content of the cell.
   */
  children: React.ReactNode;

  /**
   * The width of the cell in pixels.
   */
  width: number;

  /**
   * Optionally the height of the cell in pixels.
   * Defaults to 44.
   */
  height?: number;
};

// Props that will be injected by the EpicTable.
type InjectedProps = {
  /**
   *  Whether or not the cell is an odd cell. For the zebra stripes
   */
  odd: boolean;

  /**
   * An optional callback for when there is an action on the `EpicRow`
   * a click on one of the cells should trigger a click on the
   * `EpicRow`.
   */
  onRowClick?: (event: React.MouseEvent<HTMLDivElement>) => any;

  /**
   * Whether or not this cell should render with a hover background.
   * For the row click behavior, when hovering over one cell the
   * other cells in the row should be hovered as well. They need
   * to act as one.
   */
  hover: boolean;

  /**
   * Callback which is called whenever the hover state changes for
   * this particular cell.
   *
   * Is only called when onRowClick is defined.
   */
  onHoverChanged: (hover: boolean) => void;
};

/**
 * The EpicCell is used inside of a EpicRow to render content in.
 * It can be seen as the EpicTable's variant of the `<td>` element.
 */
export function EpicCell({ children, width, height = 44, ...rest }: Props) {
  const ref = useRef(null);

  const { odd, onRowClick, hover, onHoverChanged } = rest as InjectedProps;

  const classes = classNames('epic-table-cell border-bottom p-1', {
    'epic-table-cell--odd': odd,
    'epic-table-cell--hover': hover
  });

  function handleOnRowClick(event: React.MouseEvent<HTMLDivElement>) {
    /*
      Due to our mangeling of the DOM for the EpicTable to work. The
      normal way to prevent events from bubbling will not work via
      event.preventDefault. This is because of Reacts SyntheticEvent
      events not working like normal DOM events: 
      https://github.com/facebook/react/issues/1691

      So in order to prevent `onRowClick` from responding to events
      which came from its children, we only call `onRowClick` when
      the `div` of the `EpicCell` itself is clicked.
    */
    if (onRowClick && ref.current === event.target) {
      onRowClick(event);
    }
  }

  return (
    <div
      ref={ref}
      onMouseEnter={onRowClick ? () => onHoverChanged(true) : undefined}
      onMouseLeave={onRowClick ? () => onHoverChanged(false) : undefined}
      onClick={onRowClick ? handleOnRowClick : undefined}
      className={classes}
      style={{
        minWidth: width,
        width,
        height
      }}
    >
      {children}
    </div>
  );
}
