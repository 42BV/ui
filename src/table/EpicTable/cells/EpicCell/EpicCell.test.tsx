import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { EpicCell } from './EpicCell';

describe('Component: EpicCell', () => {
  function setup({ hasOnRowClick = true }: { hasOnRowClick?: boolean }) {
    const onRowClickSpy = jest.fn();
    const onHoverChangedSpy = jest.fn();

    const epicCell = shallow(
      <EpicCell
        width={300}
        height={44}
        // @ts-ignore
        onRowClick={hasOnRowClick ? onRowClickSpy : undefined}
        // @ts-ignore
        onHoverChanged={hasOnRowClick ? onHoverChangedSpy : undefined}
      >
        epic cell
      </EpicCell>
    );

    return { epicCell, onRowClickSpy, onHoverChangedSpy };
  }

  test('ui', () => {
    const { epicCell } = setup({});

    expect(toJson(epicCell)).toMatchSnapshot();
  });

  describe('events', () => {
    describe('onRowClick', () => {
      test(`when onRowClick exists and when clicking the EpicCell's div it should call onRowClick with the event`, () => {
        const { epicCell, onRowClickSpy } = setup({});

        // Will match the useRef initial value of null
        const fakeEvent = { target: null };

        // @ts-ignore
        epicCell
          .find('div')
          .props()
          // @ts-ignore
          .onClick(fakeEvent);

        expect(onRowClickSpy).toBeCalledTimes(1);
        expect(onRowClickSpy).toBeCalledWith(fakeEvent);
      });

      test(`when onRowClick exists and when not clicking the EpicCell's div it should not call onRowClick`, () => {
        const { epicCell, onRowClickSpy } = setup({});

        // Will not match the useRef initial value of null
        const fakeEvent = { target: 42 };

        // @ts-ignore
        epicCell
          .find('div')
          .props()
          // @ts-ignore
          .onClick(fakeEvent);

        expect(onRowClickSpy).toBeCalledTimes(0);
      });

      test(`when onRowClick does not exist and when clicking the EpicCell's div it should not call onRowClick`, () => {
        const { epicCell } = setup({ hasOnRowClick: false });

        // @ts-ignore
        const onClick =
          // @ts-ignore
          epicCell.find('div').props().onClick;

        expect(onClick).toBe(undefined);
      });
    });

    describe('hover behaviour', () => {
      it('should when onRowClick exists call onHoverChanged when hovering over the div', () => {
        const { epicCell, onHoverChangedSpy } = setup({});

        // @ts-ignore
        epicCell
          .find('div')
          .props()
          // @ts-ignore
          .onMouseEnter();

        expect(onHoverChangedSpy).toBeCalledTimes(1);
        expect(onHoverChangedSpy).toBeCalledWith(true);

        // @ts-ignore
        epicCell
          .find('div')
          .props()
          // @ts-ignore
          .onMouseLeave();

        expect(onHoverChangedSpy).toBeCalledTimes(2);
        expect(onHoverChangedSpy).toBeCalledWith(false);
      });

      it('should when onRowClick does not exist not call onHoverChanged when hovering over the div', () => {
        const { epicCell } = setup({ hasOnRowClick: false });

        const { onMouseEnter, onMouseLeave } = epicCell.find('div').props();

        expect(onMouseEnter).toBe(undefined);
        expect(onMouseLeave).toBe(undefined);
      });
    });
  });
});
