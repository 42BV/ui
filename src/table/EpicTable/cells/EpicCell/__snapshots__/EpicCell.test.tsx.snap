// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: EpicCell ui default 1`] = `
<div>
  <div
    class="epic-table-cell border-bottom p-1"
    style="min-width: 300px; width: 300px; height: 44px;"
  >
    epic cell
  </div>
</div>
`;
